File: src/main/java/main/commands/commands/misc/LyricsCommand.java
package main.commands.commands.misc;

import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo;
import lombok.SneakyThrows;
import main.audiohandlers.lavalink.LavaLinkGuildMusicManager;
import main.audiohandlers.RobertifyAudioManager;
import main.audiohandlers.sources.RobertifyAudioTrack;
import main.commands.CommandContext;
import main.commands.ICommand;
import main.utils.RobertifyEmbedUtils;
import main.utils.genius.GeniusAPI;
import main.utils.genius.GeniusSongSearch;
import me.duncte123.botcommons.messaging.EmbedUtils;
import net.dv8tion.jda.api.entities.GuildVoiceState;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Message;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.script.ScriptException;
import java.io.IOException;
import java.util.List;
import java.util.concurrent.atomic.AtomicReference;

public class LyricsCommand implements ICommand {
    private final Logger logger = LoggerFactory.getLogger(LyricsCommand.class);

    @Override @SneakyThrows
    public void handle(CommandContext ctx) throws ScriptException {
        final Member member = ctx.getMember();
        final Message msg = ctx.getMessage();
        final GuildVoiceState memberVoiceState = member.getVoiceState();
        final GuildVoiceState selfVoiceState = ctx.getSelfMember().getVoiceState();
        final var guild = ctx.getGuild();

        final List<String> args = ctx.getArgs();

        String query;

        if (args.isEmpty()) {
            if (!memberVoiceState.inVoiceChannel()) {
                msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "You must be in a voice channel to use this command")
                                .build())
                        .queue();
                return;
            }

            if (!selfVoiceState.inVoiceChannel()) {
                msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "I must be in a voice channel to use this command")
                                .build())
                        .queue();
                return;
            }

            if (!memberVoiceState.getChannel().equals(selfVoiceState.getChannel())) {
                msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "You must be in the same voice channel as I am to use this command")
                                .build())
                        .queue();
                return;
            }

            final var musicManager = RobertifyAudioManager.getInstance().getMusicManager(ctx.getGuild());
            final var audioPlayer = musicManager.getPlayer();
            final AudioTrack playingTrack = audioPlayer.getPlayingTrack();

            if (playingTrack == null) {
                msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "There is nothing playing!").build())
                        .queue();
                return;
            }

            if (!(playingTrack instanceof RobertifyAudioTrack)) {
                msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "This command is only supported by Spotify/Deezer tracks!").build())
                        .queue();
                return;
            }

            AudioTrackInfo trackInfo = playingTrack.getInfo();
            query = trackInfo.title + " by " + trackInfo.author;
        } else {
            query = String.join(" ", args);
        }

        AtomicReference<String> finalQuery = new AtomicReference<>(query);
        msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, "Now looking for: `"+query+"`").build())
                .queue(lookingMsg -> {

                    GeniusAPI geniusAPI = new GeniusAPI();
                    GeniusSongSearch songSearch;
                    try {
                        songSearch = geniusAPI.search(finalQuery.get());
                    } catch (IOException e) {
                        logger.error("[FATAL ERROR] Unexpected error!", e);
                        return;
                    }

                    if (songSearch.getStatus() == 403) {
                        lookingMsg.editMessageEmbeds(RobertifyEmbedUtils.embedMessage(guild, "I do not have permission to execute this request!").build())
                                .queue();
                        return;
                    }

                    if (songSearch.getStatus() == 404 || songSearch.getHits().size() == 0) {
                        lookingMsg.editMessageEmbeds(RobertifyEmbedUtils.embedMessage(guild, "Nothing was found for `"+ finalQuery +"`").build())
                                .queue();
                        return;
                    }

                    GeniusSongSearch.Hit hit = songSearch.getHits().get(0);

                    lookingMsg.editMessageEmbeds(RobertifyEmbedUtils.embedMessageWithTitle(guild, hit.getTitle() + " by " + hit.getArtist().getName(),
                                    hit.fetchLyrics())
                                    .setThumbnail(hit.getImageUrl())
                                    .build())
                            .queue();
                });
    }

    @Override
    public String getName() {
        return "lyrics";
    }

    @Override
    public String getHelp(String prefix) {
        return """
                Get the lyrics for the song being played

                **__Usages__**
                `lyrics` *(Fetches the lyrics for the song being currently played)*
                `lyrics <songname>` *(Fetches the lyrics for a specific song)*""";
    }

    @Override
    public List<String> getAliases() {
        return List.of("lyr");
    }
}

File: src/main/kotlin/main/commands/slashcommands/audio/LyricsCommand.kt
package main.commands.slashcommands.audio

import dev.minn.jda.ktx.coroutines.await
import dev.minn.jda.ktx.util.SLF4J
import main.audiohandlers.RobertifyAudioManager
import main.audiohandlers.utils.author
import main.audiohandlers.utils.title
import main.utils.RobertifyEmbedUtils.Companion.editEmbed
import main.utils.RobertifyEmbedUtils.Companion.replyEmbed
import main.utils.RobertifyEmbedUtils.Companion.sendEmbed
import main.utils.component.interactions.slashcommand.AbstractSlashCommand
import main.utils.component.interactions.slashcommand.models.SlashCommand
import main.utils.component.interactions.slashcommand.models.CommandOption
import main.utils.genius.GeniusAPI
import main.utils.locale.messages.GeneralMessages
import main.utils.locale.messages.LyricsMessages
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import java.io.IOException

class LyricsCommand : AbstractSlashCommand(
    SlashCommand(
        name = "lyrics",
        description = "Get the lyrics for the song being played.",
        options = listOf(
            CommandOption(
                name = "song",
                description = "The song to lookup the lyrics for",
                required = false
            )
        ),
        isPremium = true
    )
) {

    companion object {
        private val logger by SLF4J
    }

    override suspend fun handle(event: SlashCommandInteractionEvent) {
        val member = event.member!!
        val guild = event.guild!!
        val memberVoiceState = member.voiceState!!
        val selfVoiceState = guild.selfMember.voiceState!!

        val query = when (event.getOption("song")) {
            null -> {
                val acChecks = audioChannelChecks(
                    memberVoiceState,
                    selfVoiceState,
                    songMustBePlaying = true
                )
                if (acChecks != null) {
                    event.replyEmbed { acChecks }
                        .setEphemeral(true)
                        .queue()
                    return
                }

                val playingTrack = RobertifyAudioManager[guild]
                    .player
                    .playingTrack!!
                "${playingTrack.title} by ${playingTrack.author}"
            }

            else -> event.getOption("song")!!.asString
        }

        val lookingMsg = event.replyEmbed {
            embed(LyricsMessages.LYRICS_SEARCHING, Pair("{query}", query))
        }.await()

        val geniusAPI = GeniusAPI()
        val songSearch = try {
            geniusAPI.search(query)
        } catch (e: IOException) {
            logger.error("Unexpected error", e)
            return
        }

        if (songSearch.status == 403) {
            lookingMsg.editEmbed(guild) {
                embed(GeneralMessages.SELF_INSUFFICIENT_PERMS)
            }.queue()
            return
        }

        if (songSearch.status == 404 || songSearch.hits.size == 0) {
            lookingMsg.editEmbed(guild) {
                embed(LyricsMessages.LYRICS_NOTHING_FOUND, Pair("{query}", query))
            }.queue()
            return
        }

        val hit = songSearch.hits.first
        val lyrics = hit.fetchLyrics()

        if (lyrics == null) {
            lookingMsg.editEmbed(guild) {
                embed(LyricsMessages.LYRICS_NOTHING_FOUND, Pair("{query}", query))
            }.queue()
            return
        }

        try {
            lookingMsg.editEmbed(guild) {
                embed(
                    title = LyricsMessages.LYRICS_EMBED_TITLE,
                    description = lyrics,
                    Pair("{title}", hit.title),
                    Pair("{author}", hit.artist.name)
                )
            }.queue()
        } catch (e: IllegalArgumentException) {
            val chars = lyrics.length
            lookingMsg.editEmbed(guild) {
                embed(
                    LyricsMessages.LYRICS_EMBED_TITLE,
                    lyrics.substring(0, 4096),
                    Pair("{title}", hit.title),
                    Pair("{author}", hit.artist.name)
                )
            }.queue()

            var i = 4096
            do {
                event.channel.sendEmbed(guild) {
                    embed(lyrics.substring(i, chars.coerceAtMost(i + 4096)))
                }.queue()

                i += 4096
            } while (i < chars)
        }
    }

    override val help: String
        get() = "Get the lyrics for the song being played or search for the lyrics of a specific song."
}